File: bin/ruby-cli/bin/largest-files.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require_relative '../../.common/script_base'

# Script to find largest files while respecting .gitignore patterns
class LargestFilesScript < ScriptBase
  def script_emoji; '📊'; end
  def script_title; 'Largest Files Finder'; end
  def script_description; 'Find the largest files by lines (default) or size while respecting .gitignore patterns'; end
  def script_arguments; '[OPTIONS] [directory]'; end

  def add_custom_options(opts)
    opts.on('-n', '--count COUNT', Integer, 'Number of files to show (default: 20)') do |count|
      @options[:count] = count
    end
    opts.on('-s', '--size', 'Sort by file size instead of number of lines') do
      @options[:sort_by_size] = true
    end
    opts.on('-m', '--min-size SIZE', 'Minimum file size to consider (e.g., 1M, 100K)') do |size|
      @options[:min_size] = parse_size(size)
    end
    opts.on('--min-lines COUNT', Integer, 'Minimum number of lines to consider') do |count|
      @options[:min_lines] = count
    end
    opts.on('--hidden', 'Include hidden files and directories') do
      @options[:hidden] = true
    end
    opts.on('--no-gitignore', 'Ignore .gitignore patterns') do
      @options[:no_gitignore] = true
    end
  end

  def validate!
    super
    @directory = args.empty? ? original_working_dir : File.expand_path(args[0])

    unless File.directory?(@directory)
      log_error "Directory not found: #{@directory}"
      exit 1
    end

    @count = @options[:count] || 20
    @min_size = @options[:min_size] || 0
    @min_lines = @options[:min_lines] || 0
    @sort_by_size = @options[:sort_by_size] || false
    @include_hidden = @options[:hidden] || false
    @respect_gitignore = !@options[:no_gitignore]
  end

  def run
    log_banner("Finding Largest Files")

    log_info "Scanning directory: #{@directory}"
    log_info "Showing top #{@count} files" if @count > 0

    if @sort_by_size
      log_info "Sorting by: File size"
      log_info "Minimum size: #{format_size(@min_size)}" if @min_size > 0
    else
      log_info "Sorting by: Number of lines"
      log_info "Minimum lines: #{@min_lines}" if @min_lines > 0
    end

    log_info "Respecting .gitignore: #{@respect_gitignore ? 'Yes' : 'No'}"
    log_info "Including hidden files: #{@include_hidden ? 'Yes' : 'No'}"

    gitignore_patterns = load_gitignore_patterns if @respect_gitignore

    log_progress "Scanning for files..."
    files = find_files(@directory, gitignore_patterns)

    if files.empty?
      log_warning "No files found matching criteria"
      return
    end

    if @sort_by_size
      log_progress "Calculating file sizes..."
      files_with_metrics = get_file_sizes(files)
    else
      log_progress "Counting lines..."
      files_with_metrics = get_file_line_counts(files)
    end

    log_progress "Filtering and sorting..."
    filtered_files = filter_and_sort_files(files_with_metrics)

    display_results(filtered_files)
    show_completion("Largest files analysis")
  end

  private

  def parse_size(size_str)
    case size_str.downcase
    when /^(\d+)b?$/i
      $1.to_i
    when /^(\d+)k$/i
      $1.to_i * 1024
    when /^(\d+)m$/i
      $1.to_i * 1024 * 1024
    when /^(\d+)g$/i
      $1.to_i * 1024 * 1024 * 1024
    else
      log_error "Invalid size format: #{size_str}. Use formats like 100K, 1M, 1G"
      exit 1
    end
  end

  def format_size(bytes)
    units = ['B', 'KB', 'MB', 'GB', 'TB']
    size = bytes.to_f
    unit_index = 0

    while size >= 1024 && unit_index < units.length - 1
      size /= 1024
      unit_index += 1
    end

    if unit_index == 0
      "#{size.to_i}#{units[unit_index]}"
    else
      "#{size.round(1)}#{units[unit_index]}"
    end
  end

  def load_gitignore_patterns
    gitignore_path = File.join(@directory, '.gitignore')
    return [] unless File.exist?(gitignore_path)

    patterns = []
    File.readlines(gitignore_path).each do |line|
      line = line.strip
      next if line.empty? || line.start_with?('#')

      # Convert gitignore pattern to regex
      pattern = gitignore_pattern_to_regex(line)
      patterns << pattern if pattern
    end

    log_info "Loaded #{patterns.length} patterns from .gitignore"
    patterns
  end

  def gitignore_pattern_to_regex(pattern)
    # Handle negation patterns starting with !
    negated = pattern.start_with?('!')
    pattern = pattern[1..-1] if negated

    # Escape regex special characters except * and ?
    escaped = Regexp.escape(pattern)
    escaped = escaped.gsub('\\*', '.*').gsub('\\?', '.')

    # Handle directory patterns ending with /
    if escaped.end_with?('\/')
      escaped = escaped[0..-2] + '(/.*)?$'
    else
      escaped += '(/.*)?'
    end

    # Handle leading slash (absolute path)
    if escaped.start_with?('\/')
      escaped = '^' + escaped[1..-1]
    else
      escaped = '(^|/)' + escaped
    end

    begin
      regex = Regexp.new(escaped, Regexp::IGNORECASE)
      negated ? { regex: regex, negated: true } : { regex: regex, negated: false }
    rescue RegexpError => e
      log_warning "Invalid gitignore pattern: #{pattern} (#{e.message})"
      nil
    end
  end

  def file_ignored?(file_path, gitignore_patterns)
    return false if gitignore_patterns.empty?

    relative_path = file_path.sub(@directory + '/', '')

    gitignore_patterns.each do |pattern|
      if relative_path.match?(pattern[:regex])
        return !pattern[:negated]
      end
    end

    false
  end

  def find_files(directory, gitignore_patterns)
    files = []

    Dir.glob("#{directory}/**/*", File::FNM_DOTMATCH).each do |path|
      next unless File.file?(path)

      # Skip hidden files unless requested
      unless @include_hidden
        basename = File.basename(path)
        next if basename.start_with?('.')
      end

      # Skip if file is ignored by gitignore
      next if @respect_gitignore && file_ignored?(path, gitignore_patterns)

      files << path
    end

    log_info "Found #{files.length} files"
    files
  end

  def get_file_sizes(files)
    files.map do |path|
      begin
        size = File.size(path)
        { path: path, size: size, lines: 0 }
      rescue Errno::ENOENT
        # File might have been deleted during scanning
        nil
      end
    end.compact
  end

  def get_file_line_counts(files)
    files.map do |path|
      begin
        # Only count lines for text files to avoid binary files
        if text_file?(path)
          lines = File.readlines(path).count
        else
          lines = 0
        end
        { path: path, size: File.size(path), lines: lines }
      rescue Errno::ENOENT
        # File might have been deleted during scanning
        nil
      rescue => e
        # Handle encoding errors or other file reading issues
        log_debug "Could not read #{path}: #{e.message}" if debug?
        { path: path, size: File.size(path), lines: 0 }
      end
    end.compact
  end

  def text_file?(path)
    # Simple heuristic to detect text files by extension and content
    text_extensions = %w[.txt .rb .py .js .html .css .json .xml .yaml .yml .md .sh .zsh .sql .go .rs .java .c .cpp .h .hpp .php .pl .swift .kt .scala .clj .hs .ml .r .R .m .mm .swift .dart .ts .jsx .tsx .vue .svelte .elm .ex .exs .erl .hrl .nim .zig .v .vsh .jl .lua .tcl .tk .vim .vimrc .gitignore .gitattributes .editorconfig .dockerfile .makefile .cmake .rake .gemfile .package.json .requirements.txt .toml .ini .cfg .conf .config .env .env.example .env.local .env.development .env.test .env.production .log .out .err .stderr .stdout]

    extension = File.extname(path).downcase
    return true if text_extensions.include?(extension)

    # Check file content if extension is not recognized
    begin
      File.open(path, 'rb') do |file|
        # Read first 1024 bytes and check for null bytes (indicator of binary)
        chunk = file.read(1024)
        return !chunk.include?("\x00")
      end
    rescue
      false
    end
  end

  def filter_and_sort_files(files_with_metrics)
    # Filter by minimum criteria
    if @sort_by_size
      filtered = files_with_metrics.select { |file| file[:size] >= @min_size }
      # Sort by size (largest first)
      sorted = filtered.sort_by { |file| -file[:size] }
    else
      filtered = files_with_metrics.select { |file| file[:lines] >= @min_lines }
      # Sort by lines (largest first)
      sorted = filtered.sort_by { |file| -file[:lines] }
    end

    # Take top N files
    @count > 0 ? sorted.first(@count) : sorted
  end

  def display_results(files)
    if @sort_by_size
      log_section "Largest Files by Size"
    else
      log_section "Largest Files by Lines"
    end

    if files.empty?
      log_warning "No files found matching criteria"
      return
    end

    # Calculate totals
    total_size = files.sum { |file| file[:size] }
    total_lines = files.sum { |file| file[:lines] }

    if @sort_by_size
      puts "Total size of shown files: #{format_size(total_size)}"
    else
      puts "Total lines of shown files: #{total_lines.to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse}"
    end
    puts ""

    # Find the longest path for alignment
    max_path_length = [40, files.map { |f| f[:path].length }.max].min

    # Display header based on sorting method
    if @sort_by_size
      printf "%-#{max_path_length}s %10s %s\n", "File", "Size", "Path"
      puts "-" * (max_path_length + 10 + 50)

      files.each_with_index do |file, index|
        size_str = format_size(file[:size])
        relative_path = file[:path].sub(@directory + '/', '')

        # Truncate path if too long
        if relative_path.length > 50
          relative_path = "..." + relative_path[-47..-1]
        end

        printf "%-#{max_path_length}s %10s %s\n",
               File.basename(file[:path]),
               size_str,
               relative_path
      end
    else
      printf "%-#{max_path_length}s %8s %10s %s\n", "File", "Lines", "Size", "Path"
      puts "-" * (max_path_length + 8 + 10 + 50)

      files.each_with_index do |file, index|
        lines_str = file[:lines].to_s.reverse.gsub(/(\d{3})(?=\d)/, '\\1,').reverse
        size_str = format_size(file[:size])
        relative_path = file[:path].sub(@directory + '/', '')

        # Truncate path if too long
        if relative_path.length > 50
          relative_path = "..." + relative_path[-47..-1]
        end

        printf "%-#{max_path_length}s %8s %10s %s\n",
               File.basename(file[:path]),
               lines_str,
               size_str,
               relative_path
      end
    end

    puts ""
    if @sort_by_size
      log_info "Found #{files.length} files matching criteria"
    else
      log_info "Found #{files.length} files matching criteria"
    end
  end

  def show_examples
    puts "Examples:"
    puts "  #{script_name}                    # Show 20 files with most lines in current directory"
    puts "  #{script_name} -s                 # Show 20 largest files by size in current directory"
    puts "  #{script_name} /path/to/dir       # Show files with most lines in specific directory"
    puts "  #{script_name} -n 50              # Show 50 files with most lines"
    puts "  #{script_name} -s -n 50           # Show 50 largest files by size"
    puts "  #{script_name} -m 10M             # Show files larger than 10MB (requires -s)"
    puts "  #{script_name} --min-lines 100    # Show files with at least 100 lines"
    puts "  #{script_name} --hidden           # Include hidden files"
    puts "  #{script_name} --no-gitignore     # Ignore .gitignore patterns"
    puts "  #{script_name} -n 10 -s -m 5M /path # Custom options with directory"
  end
end

LargestFilesScript.execute if __FILE__ == $0

File: bin/bun-cli/src/scripts/utils/largest-files.ts
import { Script } from "../../core/decorators/Script";
import { Script as ScriptBase } from "../../core/base/Script";
import type { Context } from "../../core/types";
import * as path from "path";

interface FileMetrics {
  path: string;
  size: number;
  lines: number;
}

/**
 * Find largest files while respecting .gitignore patterns
 *
 * Features:
 * - Sort by lines (default) or file size
 * - Respects .gitignore patterns
 * - Filter by minimum size or lines
 * - Include/exclude hidden files
 * - Shows total statistics
 *
 * @example
 * largest-files                    # Show 20 files with most lines in current directory
 * largest-files -s                 # Show 20 largest files by size in current directory
 * largest-files /path/to/dir       # Show files with most lines in specific directory
 * largest-files -n 50              # Show 50 files with most lines
 * largest-files -s -n 50           # Show 50 largest files by size
 * largest-files -m 10M             # Show files larger than 10MB (requires -s)
 * largest-files --min-lines 100    # Show files with at least 100 lines
 * largest-files --hidden           # Include hidden files
 * largest-files --no-gitignore     # Ignore .gitignore patterns
 * largest-files -n 10 -s -m 5M /path # Custom options with directory
 */
@Script({
  emoji: "📊",
  tags: ["utils", "files"],
  args: {
    directory: {
      type: "string",
      position: 0,
      required: false,
      description: "Directory to analyze (defaults to current directory)",
    },
    count: {
      type: "integer",
      flag: "-n, --count",
      default: 20,
      description: "Number of files to show",
    },
    size: {
      type: "boolean",
      flag: "-s, --size",
      description: "Sort by file size instead of number of lines",
    },
    minSize: {
      type: "string",
      flag: "-m, --min-size",
      description: "Minimum file size to consider (e.g., 1M, 100K)",
    },
    minLines: {
      type: "integer",
      flag: "--min-lines",
      description: "Minimum number of lines to consider",
    },
    hidden: {
      type: "boolean",
      flag: "--hidden",
      description: "Include hidden files and directories",
    },
    noGitignore: {
      type: "boolean",
      flag: "--no-gitignore",
      description: "Ignore .gitignore patterns",
    },
  },
})
export class LargestFiles extends ScriptBase {
  private directory!: string;
  private count!: number;
  private minSizeBytes: number = 0;
  private minLines: number = 0;
  private sortBySize: boolean = false;
  private includeHidden: boolean = false;
  private respectGitignore: boolean = true;
  private gitignorePatterns: GitignorePattern[] = [];

  async run(ctx: Context): Promise<void> {
    this.logger.banner("Finding Largest Files");

    // Parse arguments
    this.directory = ctx.args.directory || process.cwd();
    this.count = ctx.args.count;
    this.sortBySize = ctx.args.size || false;
    this.includeHidden = ctx.args.hidden || false;
    this.respectGitignore = !ctx.args.noGitignore;
    this.minLines = ctx.args.minLines || 0;

    // Parse min size if provided
    if (ctx.args.minSize) {
      this.minSizeBytes = this.parseSize(ctx.args.minSize);
    }

    // Validate directory
    await this.requireDirectory(this.directory);

    this.logger.info(`Scanning directory: ${this.directory}`);
    if (this.count > 0) {
      this.logger.info(`Showing top ${this.count} files`);
    }

    if (this.sortBySize) {
      this.logger.info("Sorting by: File size");
      if (this.minSizeBytes > 0) {
        this.logger.info(`Minimum size: ${this.formatSize(this.minSizeBytes)}`);
      }
    } else {
      this.logger.info("Sorting by: Number of lines");
      if (this.minLines > 0) {
        this.logger.info(`Minimum lines: ${this.minLines}`);
      }
    }

    this.logger.info(`Respecting .gitignore: ${this.respectGitignore ? "Yes" : "No"}`);
    this.logger.info(`Including hidden files: ${this.includeHidden ? "Yes" : "No"}`);

    // Load gitignore patterns if needed
    if (this.respectGitignore) {
      this.gitignorePatterns = await this.loadGitignorePatterns();
    }

    this.logger.progress("Scanning for files...");
    const files = await this.findFiles();

    if (files.length === 0) {
      this.logger.warn("No files found matching criteria");
      return;
    }

    this.logger.progress(this.sortBySize ? "Calculating file sizes..." : "Counting lines...");
    const filesWithMetrics = await this.getFileMetrics(files);

    this.logger.progress("Filtering and sorting...");
    const filteredFiles = this.filterAndSortFiles(filesWithMetrics);

    this.displayResults(filteredFiles);
    this.logger.success("Largest files analysis complete");
  }

  private parseSize(sizeStr: string): number {
    const match = sizeStr.match(/^(\d+)([BKMG])?$/i);
    if (!match) {
      throw new Error(`Invalid size format: ${sizeStr}. Use formats like 100K, 1M, 1G`);
    }

    const value = parseInt(match[1]);
    const unit = (match[2] || "B").toUpperCase();

    const multipliers: Record<string, number> = {
      B: 1,
      K: 1024,
      M: 1024 * 1024,
      G: 1024 * 1024 * 1024,
    };

    return value * multipliers[unit];
  }

  private formatSize(bytes: number): string {
    const units = ["B", "KB", "MB", "GB", "TB"];
    let size = bytes;
    let unitIndex = 0;

    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }

    if (unitIndex === 0) {
      return `${size}${units[unitIndex]}`;
    }

    return `${size.toFixed(1)}${units[unitIndex]}`;
  }

  private async loadGitignorePatterns(): Promise<GitignorePattern[]> {
    const gitignorePath = path.join(this.directory, ".gitignore");

    if (!(await this.fs.exists(gitignorePath))) {
      return [];
    }

    const content = await this.fs.readFile(gitignorePath);
    const patterns: GitignorePattern[] = [];

    for (const line of content.split("\n")) {
      const trimmed = line.trim();
      if (!trimmed || trimmed.startsWith("#")) continue;

      const pattern = this.gitignorePatternToRegex(trimmed);
      if (pattern) {
        patterns.push(pattern);
      }
    }

    this.logger.info(`Loaded ${patterns.length} patterns from .gitignore`);
    return patterns;
  }

  private gitignorePatternToRegex(pattern: string): GitignorePattern | null {
    try {
      // Handle negation patterns starting with !
      const negated = pattern.startsWith("!");
      if (negated) {
        pattern = pattern.slice(1);
      }

      // Escape regex special characters except * and ?
      let escaped = pattern.replace(/[.+^${}()|[\]\\]/g, "\\$&");
      escaped = escaped.replace(/\*/g, ".*").replace(/\?/g, ".");

      // Handle directory patterns ending with /
      if (escaped.endsWith("/")) {
        escaped = escaped.slice(0, -1) + "(/.*)?$";
      } else {
        escaped += "(/.*)?";
      }

      // Handle leading slash (absolute path)
      if (escaped.startsWith("/")) {
        escaped = "^" + escaped.slice(1);
      } else {
        escaped = "(^|/)" + escaped;
      }

      const regex = new RegExp(escaped, "i");
      return { regex, negated };
    } catch (error) {
      this.logger.warn(`Invalid gitignore pattern: ${pattern}`);
      return null;
    }
  }

  private isFileIgnored(filePath: string): boolean {
    if (this.gitignorePatterns.length === 0) return false;

    const relativePath = path.relative(this.directory, filePath);

    for (const pattern of this.gitignorePatterns) {
      if (pattern.regex.test(relativePath)) {
        return !pattern.negated;
      }
    }

    return false;
  }

  private async findFiles(): Promise<string[]> {
    const result = await this.fs.glob({
      pattern: "**/*",
      cwd: this.directory,
      ignore: this.includeHidden ? [] : ["**/.*", "**/.*/**"],
    });

    const files: string[] = [];

    for (const file of result) {
      const fullPath = path.join(this.directory, file);

      // Check if it's a file
      if (!(await this.fs.isFile(fullPath))) continue;

      // Skip hidden files unless requested
      if (!this.includeHidden && path.basename(file).startsWith(".")) continue;

      // Skip if ignored by gitignore
      if (this.respectGitignore && this.isFileIgnored(fullPath)) continue;

      files.push(fullPath);
    }

    this.logger.info(`Found ${files.length} files`);
    return files;
  }

  private async getFileMetrics(files: string[]): Promise<FileMetrics[]> {
    const metrics: FileMetrics[] = [];

    for (const filePath of files) {
      try {
        const stats = await this.fs.stat(filePath);
        let lines = 0;

        if (!this.sortBySize) {
          // Only count lines for text files
          if (await this.isTextFile(filePath)) {
            const content = await this.fs.readFile(filePath);
            lines = content.split("\n").length;
          }
        }

        metrics.push({
          path: filePath,
          size: stats.size,
          lines,
        });
      } catch (error) {
        this.logger.debug(`Could not read ${filePath}: ${(error as Error).message}`);
      }
    }

    return metrics;
  }

  private async isTextFile(filePath: string): Promise<boolean> {
    const textExtensions = [
      ".txt", ".rb", ".py", ".js", ".html", ".css", ".json", ".xml", ".yaml", ".yml",
      ".md", ".sh", ".zsh", ".sql", ".go", ".rs", ".java", ".c", ".cpp", ".h", ".hpp",
      ".php", ".pl", ".swift", ".kt", ".scala", ".clj", ".hs", ".ml", ".r", ".R", ".m",
      ".mm", ".dart", ".ts", ".jsx", ".tsx", ".vue", ".svelte", ".elm", ".ex", ".exs",
      ".erl", ".hrl", ".nim", ".zig", ".v", ".vsh", ".jl", ".lua", ".tcl", ".tk", ".vim",
      ".vimrc", ".gitignore", ".gitattributes", ".editorconfig", ".dockerfile", ".makefile",
      ".rake", ".gemfile", ".toml", ".ini", ".cfg", ".conf", ".config", ".env", ".log",
    ];

    const ext = path.extname(filePath).toLowerCase();
    if (textExtensions.includes(ext)) return true;

    // Check file content for null bytes (indicator of binary)
    try {
      const file = Bun.file(filePath);
      const buffer = await file.arrayBuffer();
      const bytes = new Uint8Array(buffer.slice(0, 1024));

      // Check for null bytes
      for (const byte of bytes) {
        if (byte === 0) return false;
      }

      return true;
    } catch {
      return false;
    }
  }

  private filterAndSortFiles(files: FileMetrics[]): FileMetrics[] {
    // Filter by minimum criteria
    let filtered = files;

    if (this.sortBySize) {
      filtered = filtered.filter(f => f.size >= this.minSizeBytes);
      // Sort by size (largest first)
      filtered.sort((a, b) => b.size - a.size);
    } else {
      filtered = filtered.filter(f => f.lines >= this.minLines);
      // Sort by lines (largest first)
      filtered.sort((a, b) => b.lines - a.lines);
    }

    // Take top N files
    return this.count > 0 ? filtered.slice(0, this.count) : filtered;
  }

  private displayResults(files: FileMetrics[]): void {
    this.logger.section(this.sortBySize ? "Largest Files by Size" : "Largest Files by Lines");

    if (files.length === 0) {
      this.logger.warn("No files found matching criteria");
      return;
    }

    // Calculate totals
    const totalSize = files.reduce((sum, f) => sum + f.size, 0);
    const totalLines = files.reduce((sum, f) => sum + f.lines, 0);

    if (this.sortBySize) {
      console.log(`Total size of shown files: ${this.formatSize(totalSize)}`);
    } else {
      console.log(`Total lines of shown files: ${totalLines.toLocaleString()}`);
    }
    console.log();

    // Display header based on sorting method
    const maxPathLength = Math.min(40, Math.max(...files.map(f => path.basename(f.path).length)));

    if (this.sortBySize) {
      console.log(`${"File".padEnd(maxPathLength)} ${"Size".padStart(10)} Path`);
      console.log("-".repeat(maxPathLength + 10 + 50));

      for (const file of files) {
        const sizeStr = this.formatSize(file.size);
        const relativePath = path.relative(this.directory, file.path);
        const displayPath = relativePath.length > 50 ? "..." + relativePath.slice(-47) : relativePath;

        console.log(
          `${path.basename(file.path).padEnd(maxPathLength)} ${sizeStr.padStart(10)} ${displayPath}`
        );
      }
    } else {
      console.log(`${"File".padEnd(maxPathLength)} ${"Lines".padStart(8)} ${"Size".padStart(10)} Path`);
      console.log("-".repeat(maxPathLength + 8 + 10 + 50));

      for (const file of files) {
        const linesStr = file.lines.toLocaleString();
        const sizeStr = this.formatSize(file.size);
        const relativePath = path.relative(this.directory, file.path);
        const displayPath = relativePath.length > 50 ? "..." + relativePath.slice(-47) : relativePath;

        console.log(
          `${path.basename(file.path).padEnd(maxPathLength)} ${linesStr.padStart(8)} ${sizeStr.padStart(10)} ${displayPath}`
        );
      }
    }

    console.log();
    this.logger.info(`Found ${files.length} files matching criteria`);
  }
}

interface GitignorePattern {
  regex: RegExp;
  negated: boolean;
}
